### PR TITLE
Added LinearSG species

### PR DESCRIPTION
**Pull request by jmysona**

### CODE REVIEW

File: src/mcMd/species/LinearSG.cpp [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+   */
+
+   /*
+   * Read atom structure and two sets of atom type ids.
+   */
+   void LinearSG::readSpeciesParam(std::istream& in)
+   {
+      read<int>(in,"nAtom", nAtom_);
+      nBond_ = nAtom_ - 1;
+      #ifdef INTER_ANGLE
```
Comment (dmorse) [inline src/mcMd/species/LinearSG.cpp]: Need to read hasAngles_ bool variable to specify whether the molecule has an angle 
potential or not. If hasAngles_, it need to read an angleType_ integer. See HomopolymerS
as a model.

File: src/mcMd/species/LinearSG.cpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+      nAngle_ = nAtom_ - 2;
+      #endif
+      #ifdef INTER_DIHEDRAL
+      if (nAtom_ > 3)
```
Comment (dmorse) [inline src/mcMd/species/LinearSG.cpp]: Need to read hasDihedrals_ and (if hasDihedrals) dihedralType_, as in HomopolymerS class.

File: src/mcMd/species/LinearSG.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+   /*
+   * Return 0 for every bond.
+   *
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateBondTypeId(int index) const
+   { return 0; }
+
+   #ifdef INTER_ANGLE
```
Comment (dmorse) [inline src/mcMd/species/LinearSG.cpp]: Change all occurrences of INTER_ANGLE, INTER_DIHEDRAL, INTER_NOPAIR, etc. to 
SIMP_ANGLE, SIMP_DIHEDRAL, SIMP_NOPAIR, etc. Note that the preprocessor macro
names changed to reflect a change in the name of the namespace from to Simp.  This
is why the compiler can't find the functions calculateAngleTypeId and dihedralTypeId.

File: src/mcMd/species/LinearSG.cpp [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+   /*
+   * Return 0 for every angle.
+   *
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateAngleTypeId(int index) const
+   { return 0; }
+   #endif
+
+   #ifdef INTER_DIHEDRAL
```
Comment (dmorse) [inline src/mcMd/species/LinearSG.cpp]: INTER_DIHEDRAL -> SIMP_DIHEDRAL.

The function should return dihedralTypeId_

File: src/mcMd/species/LinearSG.cpp [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            beadIdentities_ = beadTypeIds1_;
+         } else {
+            beadIdentities_ = beadTypeIds2_;
+         }
+         molecule.atom(i).setTypeId(beadIdentities_[i]);
+      }
+      setMoleculeStateId(molecule, stateId);
+   }
+
+   /*
```
Comment (dmorse) [inline src/mcMd/species/LinearSG.cpp]: The current Species interface doesn't have any functions to generate random molecules. There is 
instead a separate Generator base class. Subclasses of generators are created by a GeneratorFactory,
that (for now) will instantiate a LinearGenerator for any subclass of Linear.  The only two Generator
subclasses that know exist are a PointGenerator and a LinearGenerator.

Comment (dmorse) [reply src/mcMd/species/LinearSG.cpp]: Sorry, I now see that these are commented out. Please remove commented out function declarations (in header) and definitions (here)

**Review by dmorse — CHANGES_REQUESTED** [review]
The main required changes are:
1) Mimic treatment of angles and dihedrals in HomopolymerSG, particularly in readSpeciesParam
2) Change INTER_ANGLE etc. macros to SIMP_ANGLE etc. 
3) Remove commented out generation functions

**Comment by dmorse** [comment]
Josh - I knew there was a problem because the automatic tests failed. The automatic tests are run with angles, dihedrals and external potentials enabled (-a1 -d1 -e1) so make sure you test with same flags.

**Comment by jmysona** [comment]
Dave,

I made the changes as requested. The code has passed the tests performed by
travis and restarts properly.

--Josh

**Comment by dmorse** [comment]
Josh, 
     I merged the pull request, because nothing appeared wrong, but didn't
understand why you chose to implement this function this way:

 void LinearSG::setMoleculeState(Molecule& molecule, int stateId)
  {
       int nAtom  = molecule.nAtom();
       DArray<int> beadIdentities;
       for (int i = 0; i < nAtom; ++i) {
          if (stateId == 0) {
             beadIdentities = beadTypeIds0_;
          } else {
             beadIdentities = beadTypeIds1_;
          }
          molecule.atom(i).setTypeId(beadIdentities[i]);
       }
       setMoleculeStateId(molecule, stateId);
  }

Questions:
   1) Why repeatedly make a copy of an entire array inside a loop over atoms, rather 
       than copying the array before entering the loop? 
   2) Why make a temporary local copy of the entire appropriate beadTypeIds_ array 
       at all? All you need to do is assign appropriate values to atom type Ids. Why not
      just:

       int typeId
       for (int i = 0; i < nAtom; ++i) {
            if (stateId == 0) {
                typeId = beadTypeIds0_[i];
            } else {
                typeId = beadTypeIds1_[i];
           }
           molecule.atom(i).setTypeId(beadTypeId);
      }

      If you agree, please change and send another tiny pull request.

**Comment by jmysona** [comment]
That's a very good point. The repeated copying of the array is excessive. I
think the original intention was to avoid having to evaluate an if
statement multiple times.

Currently I've gone with a third option which avoids the repeated
evaluation of the if statement; the code checks to see which subtype it
should set the molecule to, then loops over the atoms assigning them their
new types. The code is now

      int nAtom  = molecule.nAtom();
      if (stateId == 0) {
        for (int i = 0; i < nAtom; ++i) {
          molecule.atom(i).setTypeId(beadTypeIds0_[i]);
        }
      } else {
        for (int i = 0; i < nAtom; ++i) {
          molecule.atom(i).setTypeId(beadTypeIds1_[i]);
        }
      }
      setMoleculeStateId(molecule, stateId);

Let me know if this is acceptable

**Comment by dmorse** [comment]
Josh,
    That looks fine, and its also the most efficient option.
-Dave

**Pull request merged**